### PR TITLE
Add extra filter for checking if registries have changed

### PR DESCRIPTION
**Pull request by umohnani8**
A resync happens about every 20 minutes, which sends an updated
event to the image informer even if nothing in the image CR has
changed. Adding and extra filter that checks if there has been
any changes to the registries part of the image CR before syncing
the image handler again.

Helps fix https://github.com/openshift/machine-config-operator/issues/453 where changes to the image config will not be applied unless and actual change happened in the CR.

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
@runcom @mrunalp @kikisdeliveryservice PTAL

**Comment by kikisdeliveryservice** [comment]
FTR: Lots of CI flakes with 504 gateway timeouts hence the retests.

Taking a look/testing locally.

**Comment by kikisdeliveryservice** [comment]
@umohnani8 : is there something I should be testing for other than watching the logs? If so, lmk!

**Comment by runcom** [comment]
/hold

discussed offline, we'll need something lighter and more state oriented

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go [hunk 1]
```diff
@@ -530,21 +531,24 @@ func (ctrl *Controller) syncImageConfig(key string) error {
 	// Deep-copy otherwise we are mutating our cache.
 	imgcfg = imgcfg.DeepCopy()
 
+	// Check that a change to the registries part in the image CR actually happened
+	// if isNewData := ctrl.newRegistriesData(imgcfg.Spec); !isNewData {
+	// 	// There was no change, so no need to run through the sync again
+	// 	return nil
+	// }
```
Comment (runcom) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go]: Drop this

Comment (umohnani8) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go]: woops, dropping.

**Comment by umohnani8** [comment]
@runcom updated code as discussed on slack.
@kikisdeliveryservice yeah you shouldn't see "Applied Image..." in the logs unless you actually go update the image cr (oc edit image.config.openshift.io cluster)

**Comment by kikisdeliveryservice** [comment]
hi @umohnani8 since the docs are pretty thin on using this feature and im less familiar with it, can you please add step by step instructions on how to test this?

**Comment by umohnani8** [comment]
@kikisdeliveryservice this is what an example image CR looks like with insecure and blocked registries in it

```
apiVersion: config.openshift.io/v1
kind: Image
metadata:
  creationTimestamp: 2019-02-19T16:01:32Z
  generation: 6
  name: cluster
  resourceVersion: "282342"
  selfLink: /apis/config.openshift.io/v1/images/cluster
  uid: 9ff69432-345f-11e9-8ea5-0a4bee075008
spec:
  additionalTrustedCA:
    name: ""
  registrySources:
    insecureRegistries:
    - blah.io
    blockedRegistries:
    - test.io
status:
  internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
```

You can add as many or as little you want there and should see the changes on all the nodes once the controller does its magic :)
The CR usually exists already in a new cluster (so I have noticed with all my clusters) so all you need to do is `oc edit image.config.openshift.io cluster`. Since it is a cluster wide config, we only care for the CR called "cluster".

**Comment by umohnani8** [comment]
Looks like the test flakes being fixed in #457 
/test unit

**Comment by kikisdeliveryservice** [comment]
thanks @umohnani8 :) will try this out.

File: pkg/controller/container-runtime-config/container_runtime_config_controller.go [hunk 2]
```diff
@@ -536,15 +537,12 @@ func (ctrl *Controller) syncImageConfig(key string) error {
 		return err
 	}
 	for _, pool := range mcpPools {
+		// To keep track of whether we "actually" got an updated image config
+		updated := true
```
Comment (runcom) [inline pkg/controller/container-runtime-config/container_runtime_config_controller.go]: Let's call this `applied`

Comment (umohnani8) [reply pkg/controller/container-runtime-config/container_runtime_config_controller.go]: done

**Comment by kikisdeliveryservice** [comment]
Testing results:
1. in a cluster running for a while, no more erroneous logging/syncing
2. editing `oc edit image.config.openshift.io cluster` and adding insecure and secure registries:
```
I0220 00:47:17.343506       1 container_runtime_config_controller.go:600] Applied ImageConfig cluster on MachineConfigPool worker
I0220 00:47:18.649066       1 container_runtime_config_controller.go:600] Applied ImageConfig cluster on MachineConfigPool master
I0220 00:47:22.941391       1 render_controller.go:456] Generated machineconfig worker-cac98dc94bdf965eb43a5bddfb46eb6e from 5 configs: [{MachineConfig  00-worker  machineconfiguration.openshift.io/v1  } {MachineConfig  00-worker-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-worker-container-runtime  machineconfiguration.openshift.io/v1  } {MachineConfig  01-worker-kubelet  machineconfiguration.openshift.io/v1  } {MachineConfig  99-worker-9446842e-33cd-11e9-8221-02427fa2e0b0-registries  machineconfiguration.openshift.io/v1  }]
I0220 00:47:24.241613       1 render_controller.go:456] Generated machineconfig master-3745275ffcaf2c65dec35fc28e164976 from 5 configs: [{MachineConfig  00-master  machineconfiguration.openshift.io/v1  } {MachineConfig  00-master-ssh  machineconfiguration.openshift.io/v1  } {MachineConfig  01-master-container-runtime  machineconfiguration.openshift.io/v1  } {MachineConfig  01-master-kubelet  machineconfiguration.openshift.io/v1  } {MachineConfig  99-master-94454f94-33cd-11e9-8221-02427fa2e0b0-registries  machineconfiguration.openshift.io/v1  }]
I0220 00:47:28.041363       1 node_controller.go:432] Setting node ip-10-0-138-64.us-west-2.compute.internal to desired config worker-cac98dc94bdf965eb43a5bddfb46eb6e
I0220 00:47:29.441259       1 node_controller.go:432] Setting node ip-10-0-150-226.us-west-2.compute.internal to desired config master-3745275ffcaf2c65dec35fc28e164976
I0220 00:48:46.870972       1 node_controller.go:432] Setting node ip-10-0-160-154.us-west-2.compute.internal to desired config master-3745275ffcaf2c65dec35fc28e164976
```
Checking MCD: 
```
I0220 00:52:50.479085    4777 update.go:137] Checking reconcilable for config worker-f2f49ab8040cc5907aa3d81afa0be316 to worker-cac98dc94bdf965eb43a5bddfb46eb6e
...
I0220 00:52:51.972143    4777 update.go:663] machine-config-daemon initiating reboot: Node will reboot into config worker-cac98dc94bdf965eb43a5bddfb46eb6e
```

new config was successfully applied and daemons rebooted.

**Comment by kikisdeliveryservice** [comment]
Flakes:  `The bootstrap user should successfully login with password decoded from kubeadmin secret` & cluster network timeouts.

/retest

**Comment by runcom** [comment]
CI errors tracked in slack

/retest

**Pull request merged**